src/ORM/Behavior/CounterCacheBehavior.php

```diff
@@ -190,6 +192,85 @@ public function afterDelete(EventInterface $event, EntityInterface $entity, Arra
         $this->_processAssociations($event, $entity);
     }
 
+    /**
+     * Update counter cache.
+     *
+     * @param string|null $assocName The association name to update counter cache for.
+     *  If null, all configured associations will be updated.
+     * @throws \InvalidArgumentException If specified association is not configured.
+     * @return void
+     */
+    public function updateCounterCache(?string $assocName = null): void
+    {
+        $config = $this->_config;
+        if ($assocName !== null) {
+            if (!isset($config[$assocName])) {
+                throw new InvalidArgumentException(sprintf(
+                    'Association `%s` is not configured for counter cache behavior of `%s`',
+                    $assocName,
+                    $this->_table->getAlias()
+                ));
+            }
+
+            $config = [$assocName => $config[$assocName]];
+        }
+
+        foreach ($config as $assoc => $settings) {
+            /** @var \Cake\ORM\Association\BelongsTo $belongsTo */
+            $belongsTo = $this->_table->getAssociation($assoc);
+
+            foreach ($settings as $field => $config) {
+                if ($config instanceof Closure) {
+                    // Cannot update counter cache which use a closure
+                    return;
+                }
+
+                if (is_int($field)) {
+                    $field = $config;
+                    $config = [];
+                }
+
+                $this->updateCountForAssociation($belongsTo, $field, $config);
+            }
+        }
+    }
+
+    /**
+     * Update counter cache for the given association.
+     *
+     * @param \Cake\ORM\Association\BelongsTo $assoc The association object.
+     * @param string $field Counter cache field.
+     * @param array $config Config array.
+     * @return void
+     */
+    protected function updateCountForAssociation(BelongsTo $assoc, string $field, array $config): void
+    {
+        $primaryKeys = (array)$assoc->getBindingKey();
+        /** @var list<string> $foreignKeys */
+        $foreignKeys = (array)$assoc->getForeignKey();
+
+        $results = $assoc->getTarget()->find()
+            ->select($primaryKeys)
+            ->all();
```

Comment:
Can we have a limit/offset on this? Eventually we'll OOM here and if we give limit/offset applications can iteratively page their way through a large dataset without needing a ton of RAM.

Reply:
Makes sense, though I am wondering if updating each record individually is a good idea. After the sub query changes, for each records this would generate queries like 
```
update users set post_count = (select count(*) from posts where user_id = X) where users.id = X
``` 
when all the records could be be updated in a single query like 
```
update users left join posts on users.id=user_id set post_count = (select count(*) from posts where user_id = users.id)
```

Though the latter might lock up the whole table?

Reply:
> Though the latter might lock up the whole table?

I don't remember the details of how MySQL/Maria take locks for updates. But I do know that updates with JOIN don't exist in postgres which complicates doing that kind of query in a cross vendor way.

Reply:
> ..updates with JOIN don't exist in postgres..

That settles it then, we stick with what I have implemented currently.